src/world/world_generator.rs

```diff
@@ -0,0 +1,174 @@
+    // odd number, determines smoothness of the final surface, higher numbers will make the surface smoother
+    const FILTER_WIDTH: u32 = 3;
+    // number of random height points including map edges
+    const HEIGHT_POINT_COUNT: u32 = 2 + MAP_WIDTH / 150;
+
+    let mut surface_height_map: [u32; MAP_WIDTH as usize] = [0; MAP_WIDTH as usize];
+
+    let mut random_height_points: [u32; HEIGHT_POINT_COUNT as usize] = [0; HEIGHT_POINT_COUNT as usize];
+    for i in 0..HEIGHT_POINT_COUNT {
+        random_height_points[i as usize] = rng.gen_range(HEIGHT_POINT_RANGE);
+    }
+
+    fn filter(x: f32, offset: f32) -> f32 {
+        if x == offset {
+            1.5
+        } else {
+            1.5 / (std::f32::consts::PI * (x - offset)) * f32::sin(std::f32::consts::PI * (x - offset))
+        }
+    }
+
+    for x in 0..MAP_WIDTH {
+        let mapped_x: f32 = x as f32 / (MAP_WIDTH - 1) as f32 * (HEIGHT_POINT_COUNT - 1) as f32;
+        let interpolated_y: f32 = random_height_points[mapped_x.floor() as usize] as f32 * filter(
+            mapped_x.floor(),
+            mapped_x,
+        );
+
+        let mut filter_value: f32 = 0.0;
+
+        for filter_x in mapped_x.floor() as u32 + 1..mapped_x.floor() as u32 + 1 + (FILTER_WIDTH - 1) / 2 {
+            if filter_x < HEIGHT_POINT_COUNT {
+                filter_value += random_height_points[filter_x as usize] as f32 * filter(
+                    filter_x as f32,
+                    mapped_x,
+                );
+            }
+        }
+        for filter_x in mapped_x.floor() as i32 - 1 - (FILTER_WIDTH - 1) as i32 / 2..mapped_x.floor() as i32 - 1 {
+            if filter_x >= 0 {
+                filter_value += random_height_points[filter_x as usize] as f32 * filter(
+                    filter_x as f32,
+                    mapped_x,
+                );
+            }
+        }
+
+        let surface_y: u32 = (interpolated_y + filter_value).round() as u32; 
+
+        surface_height_map[x as usize] = surface_y;
+
+        set_block(&mut world_map, x, surface_y, BlockTypes::Grass);
+
+        for y in surface_y + 1..surface_y + 6 {
+            set_block(&mut world_map, x, y, BlockTypes::Dirt);
+        }
+        for y in surface_y + 6..MAP_HEIGHT {
+            set_block(&mut world_map, x, y, BlockTypes::Stone);
+        }
+    }
+    
+    // trees
+    const TREE_HEIGHT_RANGE: std::ops::RangeInclusive<u32> = 5..=10;
+
+    let tree_count: u32 = MAP_WIDTH / 20; // number of attempts to create a tree, 1 tree for every 20 blocks
+    for _ in 0..tree_count {
+        let tree_x: u32 = rng.gen_range(0..=MAP_WIDTH);
+        attempt_generate_surface_tree(&mut world_map, tree_x, &mut rng);
+    }
+
+    fn get_ground_y_if_clear(world_map: &Box<[u16; MAP_AREA as usize]>, x: u32) -> Option<u32> {
+        for y in 0..MAP_HEIGHT {
+            let target_block = get_block(&world_map, x, y);
+
+            match target_block {
```

Comment:
Clippy will proabbly remind you, but ? or `target_block.ok()?` (to convert `Result<T, E>` to `Option<T>`, which can be used directly with ? to return from the current function) would make this a bit cleaner